File: JavaScript/Programmers/Weekly Challenge/2주차.js
// 문제 조건에 따라 구현
// 행과 열을 바꿔서 계산하는 부분이 관건

// function grade(score) { // 학점 계산하는 함수
//     switch(Math.trunc(score / 10)) { // 학점을 계산할 때 10으로 나눈 몫으로 구함
//         case 10:
//         case 9: return "A";
//         case 8: return "B";
//         case 7: return "C";
//         case 6:
//         case 5: return "D";
//         default: return "F";
//     }
// }

// function solution(scores) {
//     const transpose = scores.reduce( // reduce 메소드를 이용하여 행과 열을 바꿔서 return
//         (result, row)=> row.map((_, idx)=> [...(result[idx] || []), row[idx]]) // result[idx] 값이 있다면 result[idx]에 row[idx]를 연결시킴
//                                                                                // result[idx] 값이 없다면 빈 배열에 row[idx]를 연결시킴
//     , []); // 초기값 빈 배열로 설정, reduce 메소드를 사용할 때 항상 초기값 설정해줄 것
//     return transpose.reduce((answer, current, idx)=> { // reduce 메소드를 이용하여 행과 열을 바꾼 배열에서 학점을 도출
//         let len = 0
//         const total = current.reduce((sum, score)=> { // 현재 행의 점수를 구함
//             // 현재 점수가 본인 점수이고, 유일한 최고점 또는 최저점이라면, 이때 filter 메소드를 이용
//             if (score === current[idx] && (score === Math.max(...current) || score === Math.min(...current)) && current.filter(a=> a === score).length === 1) {
//                 return sum; // 본인 점수 더하지 않음
//             }
//             // 조건이 아니라면
//             len++; // 학생 수 count
//             return sum + score; // 현재 점수 더함
//         }, 0);
        
//         return answer + grade(Math.trunc(total / len)); // Math.trunc를 사용하여 평균의 소수점을 잘라내고 학점을 구함 
//     }, "");
// }

// 학점을 계산하는 함수를 따로 만들지 않고, "FFFFFDDCBAA" 문자열을 이용하여 학점을 구함

// function solution(scores) {
//     const transpose = scores.reduce(
//         (result, row)=> row.map((_, idx)=> [...(result[idx] || []), row[idx]])
//     , []);
//     return transpose.reduce((answer, current, idx)=> {
//         let len = 0
//         const total = current.reduce((sum, score)=> {
//             if (score === current[idx] && (score === Math.max(...current) || score === Math.min(...current)) && current.filter(a=> a === score).length === 1) {
//                 return sum;
//             }
            
//             len++;
//             return sum + score;
//         }, 0);
        
//         return answer + "FFFFFDDCBAA"[Math.trunc(total / len / 10)]; // 평균을 10으로 나누고 소수점을 잘라내어 학점을 구함
//     }, "");
// }

// reduce 메소드 대신 map 메소드를 이용하여 매핑

function solution(scores) {
    return scores[0].map((_, c) => scores.map(r => r[c])) // 행과 열을 바꿈, 행에서 열에 해당하는 숫자를 뽑아 배열로 만듦, 2차원 배열이므로 map 메소드도 2개 있어야함
                                                          // scores[0]가 [1, 2, 3, 4, 5]라면 각 요소에 행과 열이 바뀐 배열이 들어옴
                                                          // [[1, 2, 3, 4, 5], [1, 2, 3, 4, 5], [1, 2, 3, 4, 5], [1, 2, 3, 4, 5], [1, 2, 3, 4, 5]] 이런식으로 매핑
        .map((s, i) => [...s.splice(i, 1), s]) // 본인 점수를 잘라냄 즉, [1, [2, 3, 4, 5]] 이런식으로 만듦
        .map(([m, s]) => Math.min(...s) <= m && m <= Math.max(...s) ? [m, ...s] : s) // 본인의 점수가 유일한 최고점 또는 최저점이 아니라면 본인 점수와 배열 s를 합침
                                                                                     // 본인의 점수가 유일한 최고점 또는 최저점이라면 s만 return
        .map(s => "FDDCBAA"[Math.max(parseInt(s.reduce((a, c) => a + c) / s.length / 10) - 4, 0)]) // reduce 메소드를 이용하여 평균을 구하고 10으로 나누어 소수점을 잘라냄
                                                                                                   // 문자열의 길이가 10이 아닌 6이므로 나온 값에 -4를 하고, 0과 비교하여
                                                                                                   // 최대값에 해당하는 학점을 구함
        .join(""); // 결과는 배열이므로 join 메소드를 이용하여 문자열로 합침
}

File: Python/Programmers/Weekly Challenge/2주차.py
# 문제에 나온대로 구현
# 평균을 구할 때 행 단위로 평균을 구하는 것이 아니라 열 단위로 평균을 구해야함
# 따라서 행과 열을 바꿔 리스트를 생성하고 본인 점수가 유일한 최고점 또는 최저점이라면 제외하고 평균을 구함

# def grade(score): # 각 점수에 따른 학점 함수
#     if score >= 90: return "A"
#     elif 80 <= score < 90: return "B"
#     elif 70 <= score < 80: return "C"
#     elif 50 <= score < 70: return "D"
#     else: return "F"

# def solution(scores):
#     answer = ""
#     for i in range(len(scores[0])):
#         tmp = [score[i] for score in scores] # 열을 기준으로 리스트 생성
#         if (tmp[i] == max(tmp) or tmp[i] == min(tmp)) and tmp.count(tmp[i]) == 1: # 본인 점수가 최고점 또는 최저점이고, 유일하면
#             avg = (sum(tmp) - tmp[i]) / (len(tmp) - 1) # 본인 점수를 제외하고 평균을 구함
#             answer += grade(avg) # 평균에 따른 학점을 answer에 더함
#             continue
            
#         avg = sum(tmp) / len(tmp) # 아니라면 전체 평균을 구함
#         answer += grade(avg) # 평균에 따른 학점을 answer에 더함
        
#     return answer

# 학점을 계산할 때 if문 대신 dict으로 순회를 돌면서 계산하는 걸로 수정

# def grade(score): # 학점 계산 함수
#     rank = {90: "A", 80: "B", 70: "C", 50: "D"} # 점수에 따른 학점 dict
#     for key, value in rank.items(): # rank를 순회하면서 score가 key보다 크거나 같으면 key에 해당하는 value 반환
#         if score >= key:
#             return value
        
#     return "F" # rank에 점수가 없으면 F 반환

# def solution(scores):
#     answer = ""
#     for i in range(len(scores[0])):
#         tmp = [score[i] for score in scores]
#         if (tmp[i] == max(tmp) or tmp[i] == min(tmp)) and tmp.count(tmp[i]) == 1:
#             avg = (sum(tmp) - tmp[i]) / (len(tmp) - 1)
#             answer += grade(avg)
#             continue
            
#         avg = sum(tmp) / len(tmp)
#         answer += grade(avg)

#     return answer

# 학점을 계산할 때 함수를 따로 선언하지 않고, "FFFFFDDCBAA" 문자열을 이용하여 계산

def solution(scores):
    answer = ""
    for i in range(len(scores[0])):
        tmp = [score[i] for score in scores]
        if (tmp[i] == max(tmp) or tmp[i] == min(tmp)) and tmp.count(tmp[i]) == 1: # 본인의 점수가 유일한 최고점 또는 최저점이라면
            avg = (sum(tmp) - tmp[i]) / (len(tmp) - 1) # 본인 점수를 제외하고 평균 계산
        else: # 아니라면
            avg = sum(tmp) / len(tmp) # 전체 평균 계산
            
        answer += "FFFFFDDCBAA"[int(avg) // 10] # 학점 계산, 문자열이 길이가 10이므로 평균의 소수 부분을 자르고 10으로 나눈 몫으로 학점을 찾음
        
    return answer
